### ISSUE TITLE
Docker in Docker

### ISSUE THREAD

**Issue by maldag** [readthedocs/readthedocs.org#6094]
## Details
I'm setting up a local build environment with readthedocs and like to wrap it up in an docker-compose file. There are several docker-compose files on github, but I don't know if readthedocs.org is made to use ENABLE_DOCKER for builds, when readthedocs itself is already being run in a docker environment.

Is this something you've thought about using the Docker-Py Api?

**Comment by humitos** [comment]
Hi! There were different attempts for this with different implementations but none of them currently works. There is no documentation about these attempt either.

Unfortunately, this is not officially supported and there is not too much advice that we can give you :disappointed:

**Comment by maldag** [comment]
I'd like to give you an update on this topic since I've got it working with docker in docker.

- I'm using RTD version 3.7.2
- Docker in Docker seems to be popular using jenkins as a build system, so I found some help searching for `Docker in Docker jenkins`

Anyways:

1. Run RTD via supervisord in docker as root. Setup supervisord to run the python process RTD as user docs:docs
2. The Docker API in Python creates containers when using `DOCKER_ENABLE` and will run the next container as user root, since no user is specified.
3. I patched this behaviour to be able to specify the user the build-Container will run in.
```
diff --git a/readthedocs/doc_builder/environments.py b/readthedocs/doc_builder/environments.py
index 6b6f83ddc..df2e4bc72 100644
--- a/readthedocs/doc_builder/environments.py
+++ b/readthedocs/doc_builder/environments.py
@@ -1042,6 +1042,7 @@ class DockerBuildEnvironment(BuildEnvironment):
                 host_config=self.get_container_host_config(),
                 detach=True,
                 environment=self.environment,
+                user='docs',
             )
             client.start(container=self.container_id)
         except ConnectionError:
```
4. Since docker in docker does use the same docker daemon running on the host, the volumes that link the build in- and outputs must be the same and absolute paths. That's one major drawback I haven't solved yet. So create a folder `/app` on the host and attach it to the RTD container. The build environment will then bind-mount the `/app/user_build` folder to the build image.

Once I'm done with this I might add a merge request allowing the configuration to generically start the build image docker as a specified user.
